File: user-service/rabbitmq/publisher.go
package rabbitmq

import (
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

type RabbitMQPublisher struct {
	connectionManager *RabbitMQConnectionManager
	exchange          string
	routingKey 	  string
	channel           *amqp.Channel
}

// NewRabbitMQPublisher creates a new RabbitMQPublisher instance
func NewRabbitMQPublisher(connectionManager *RabbitMQConnectionManager, exchange string) *RabbitMQPublisher {
	return &RabbitMQPublisher{
		connectionManager: connectionManager,
		exchange:          exchange,
	}
}

// Publish sends a message to the exchange. It ensures the channel is open and publishes the message.
func (r *RabbitMQPublisher) Publish(message []byte, routingKey string) error {
	var err error
	if r.channel == nil || r.channel.IsClosed() {
		r.channel = r.connectionManager.GetChannel()
	}

	// Declare the exchange
	err = r.channel.ExchangeDeclare(
		r.exchange,   // exchange name
		"topic",     // exchange type
		true,         // durable
		false,        // auto-deleted
		false,        // internal
		false,        // no-wait
		nil,          // arguments
	)
	if err != nil {
		log.Printf("Failed to declare exchange: %v", err)
		return err
	}

	// Publish the message
	err = r.channel.Publish(
		r.exchange, // exchange
		routingKey,         // routing key (empty for fanout)
		false,      // mandatory
		false,      // immediate
		amqp.Publishing{
			ContentType: "application/json",
			Body:        message,
			DeliveryMode: amqp.Persistent,
		},
	)
	if err != nil {
		log.Printf("Failed to publish message: %v", err)
		// Attempt to reconnect and republish the message
		r.channel = r.connectionManager.GetChannel()
		return r.Publish(message, routingKey) // Retry publishing
	}

	log.Printf(" [x] Sent message: %s", string(message))
	return nil
}

File: subscription-service/app/publisher.py
import logging
import pika
from connection_manager import RabbitMQConnectionManager

logger = logging.getLogger(__name__)

class RabbitMQPublisher:
    def __init__(self, connection_manager: RabbitMQConnectionManager, exchange):
        self.connection_manager = connection_manager
        self.exchange = exchange
        self.channel = None
    
    def publish(self, message, routing_key):
        try:
            # Ensure channel is open
            if self.channel is None or self.channel.is_closed:
                self.channel = self.connection_manager.get_channel()
            
            # Declare the exchange
            self.channel.exchange_declare(exchange=self.exchange, exchange_type='topic', durable=True)
            
            # Publish the message
            self.channel.basic_publish(
                exchange=self.exchange,
                routing_key=routing_key,
                body=message,
                properties=pika.BasicProperties(
                    content_type='application/json',
                    delivery_mode=pika.spec.PERSISTENT_DELIVERY_MODE
                )
            )
            logger.info(f" [x] Sent message: {message}")
        
        except pika.exceptions.AMQPError as e:
            logger.error(f"Failed to publish message due to AMQP error: {e}")
            # Try reconnecting and republishing the message
            self.channel = self.connection_manager.get_channel()
            self.publish(message, routing_key)